### PR TITLE
[Testing and Infra] LIT testing and passes enabling

### PR DESCRIPTION
**Pull request by Devjiu**
This commit enables LIT tests and introduces example of MLP LIT test.
To run all of them use:
```sh
cmake --build . --target gc-check
```

Also introduced infrastructure for adding Passes and registered dummy TileNamedLinalg pass.
Currently it simply print matmuls. To enable use option: "--tile-named-linalg".

Signed-off-by: Dmitrii Makarenko <dmitrii.makarenko@intel.com>

### CODE REVIEW

File: scripts/compile.sh [hunk 1]
```diff
@@ -22,4 +22,5 @@ get_llvm() (
 test -f "$llvm_dir/llvm-$llvm_hash"/llvm.tgz || get_llvm
 
 MLIR_DIR="$llvm_dir/lib/cmake/mlir" cmake -S . -G Ninja -B build
+# cmake -G Ninja .. -DCMAKE_BUILD_TYPE=RelWithDebInfo -DMLIR_DIR=$CUSTOM_LLVM_ROOT/lib/cmake/mlir -DLLVM_EXTERNAL_LIT=$CUSTOM_LLVM_ROOT/bin/llvm-lit -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++
```
Comment (kurapov-peter) [inline scripts/compile.sh:25]: The end goal is to be able to use the script locally. Feel free to improve it directly. Also, currently the build uses gcc.
```suggestion
```

File: test/lit.cfg.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# test_source_root: The root path where tests are located.
+config.test_source_root = os.path.dirname(__file__)
+
+# test_exec_root: The root path where tests should be run.
+config.test_exec_root = os.path.join(config.gc_obj_root, "test")
+
+config.substitutions.append(("%PATH%", config.environment["PATH"]))
+config.substitutions.append(("%shlibext", config.llvm_shlib_ext))
+config.substitutions.append(("%llvmlibdir", config.llvm_lib_dir))
+config.substitutions.append(("%tpplibdir", config.gc_obj_root + "/lib/"))
```
Comment (kurapov-peter) [inline test/lit.cfg.py:31]:
```suggestion
config.substitutions.append(("%gclibdir", config.gc_obj_root + "/lib/"))
```

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        FileCheck count not
+        # mlir-gen
+        gc-opt
+        )
+
+add_lit_testsuite(gc-check "Running the regression tests"
+        ${CMAKE_CURRENT_BINARY_DIR}
+        DEPENDS ${GC_OPT_TEST_DEPENDS}
+        )
+set_target_properties(gc-check PROPERTIES FOLDER "Tests")
```
Comment (kurapov-peter) [inline test/CMakeLists.txt:18]: What is "Tests" pointing to?

**Review by kurapov-peter — CHANGES_REQUESTED** [review]
Looks good, just a couple tweaks.

File: test/CMakeLists.txt (hunk 3 above)
Comment (Devjiu) [reply test/CMakeLists.txt:34]: In some cases written that it used for IDE. I am not sure what it does, added TODO for this.

File: scripts/compile.sh (hunk 1 above)
Comment (Devjiu) [reply scripts/compile.sh]: done

File: test/lit.cfg.py (hunk 2 above)
Comment (Devjiu) [reply test/lit.cfg.py]: done

File: CMakeLists.txt [hunk 4]
```diff
@@ -55,4 +57,3 @@ set(GC_LIB_LINKED_LIBS
 )
 target_link_libraries(graph_compiler PRIVATE ${GC_LIB_LINKED_LIBS})
 
-add_subdirectory(test)
```
Comment (kurapov-peter) [inline CMakeLists.txt:58]: What is this for?

File: lib/gc-dialects/Transforms/TileNamed.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    llvm::SmallVector<Operation *> to_tile;
+    for (Operation &o : getOperation()->getRegion(0).front().getOperations()) {
+      if (isa<linalg::MatmulOp>(o)) {
+        to_tile.push_back(&o);
+      }
+    }
+
+    for (Operation *o : to_tile) {
+      llvm::errs() << "func op body to tile: " << *o << "\n";
+      //   tryTileUsingSCF(rewriter, cast<TilingInterface>(o), tilingOptions);
```
Comment (kurapov-peter) [inline lib/gc-dialects/Transforms/TileNamed.cpp:46]: Is this what we leave for the future pass implementation?

**Review by kurapov-peter — APPROVED** [review]
Good overall, can benefit from cleanup.

File: CMakeLists.txt (hunk 4 above)
Comment (Devjiu) [reply CMakeLists.txt:58]: just moved it up. This folder contains gtests and lits.

File: lib/gc-dialects/Transforms/TileNamed.cpp (hunk 5 above)
Comment (Devjiu) [reply lib/gc-dialects/Transforms/TileNamed.cpp:46]: yep, I can remove it if it is not needed now

**Pull request merged**